flocker/provision/_ssh/_conch.py

```diff
@@ -0,0 +1,183 @@
+
+from ._model import Run, Sudo, Put, Comment, RunRemotely
+
+from .._effect import dispatcher as base_dispatcher
+
+
+logger = Logger()
+
+
+@attributes([
+    "deferred",
+    "context",
+])
+class CommandProtocol(LineOnlyReceiver, object):
+    """
+    Protocol that logs the lines of a remote command.
+
+    :ivar Deferred deferred: Deferred to fire when the command finishes
+        If the command finished successfully, will fire with ``None``.
+        Otherwise, errbacks with the reason.
+    :ivar Message context: The eliot message context to log.
+    """
+    delimiter = b'\n'
+
+    def connectionMade(self):
+        self.transport.disconnecting = False
+
+    def connectionLost(self, reason):
+        if reason.check(ConnectionDone):
+            self.deferred.callback(None)
+        else:
+            self.deferred.errback(reason)
+
+    def lineReceived(self, line):
+        self.context.bind(
+            message_type="flocker.provision.ssh:run:output",
+            line=line,
+        ).write(logger)
+
+
+@sync_performer
+def perform_sudo(dispatcher, intent):
+    """
+    See :py:class:`Sudo`.
+    """
+    return Effect(Run(command='sudo ' + intent.command))
+
+
+@sync_performer
+def perform_put(dispatcher, intent):
+    """
+    See :py:class:`Put`.
+    """
+    return Effect(Run(command='printf %s > %s'
+                              % (shell_quote(intent.content),
+                                 shell_quote(intent.path))))
+
+
+@sync_performer
+def perform_comment(dispatcher, intent):
+    """
+    See :py:class:`Comment`.
+    """
+
+
+def get_ssh_dispatcher(connection, context):
+    """
+    :ivar Message context: The eliot message context to log.
+    :ivar connection: The SSH connection run commands on.
+    """
+
+    @deferred_performer
+    def perform_run(dispatcher, intent):
+        context.bind(
+            message_type="flocker.provision.ssh:run",
+            command=intent.command,
+        ).write(logger)
+        endpoint = SSHCommandClientEndpoint.existingConnection(
+            connection, intent.command)
+        d = Deferred()
+        connectProtocol(endpoint, CommandProtocol(
+            deferred=d, context=context))
+        return d
+
+    return TypeDispatcher({
+        Run: perform_run,
+        Sudo: perform_sudo,
+        Put: perform_put,
+        Comment: perform_comment,
+    })
+
+
+def get_connection_helper(address, username, port):
+    """
+    Get a :class:`twisted.conch.endpoints._ISSHConnectionCreator` to connect to
+    the given remote.
+
+    :param bytes address: The address of the remote host to connect to.
+    :param bytes username: The user to connect as.
+    :param int port: The port of the ssh server to connect to.
+
+    :return _ISSHConnectionCreator:
+    """
+    try:
+        agentEndpoint = UNIXClientEndpoint(
+            reactor, os.environ["SSH_AUTH_SOCK"])
+    except KeyError:
+        agentEndpoint = None
+
+    return _NewConnectionHelper(
```

Comment:
Assuming that "yet" actually works in that sentence bothers me. :smile: 

The extra process is entirely harmless.  Using a private API from another project is potentially harmful.